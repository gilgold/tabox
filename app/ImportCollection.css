.image-upload {
	display: block;
	margin: 8px 0 6px 0;
	height: 30px;
    line-height: 16px;
}
.image-upload > input {
	visibility:hidden;
	width:0;
	height:0
}
.image-upload > label {
	cursor: pointer;
}
.input-label {
	margin-top: 3px;
	display: flex;
    align-items: center;
	width: 160px;
}
.import-icon {
	float: left;
}
.import-button {
	background-color: var(--primary-color);
	color: #fff;
	border: none;
	border-radius: 5px;
	padding: 6px 14px;
	display: flex;
	flex-direction: row;
	align-items: center;
	justify-content: space-between;
	font-size:.75em;
	font-weight: 600;
	flex-basis: 160px;
}