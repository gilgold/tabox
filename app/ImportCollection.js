import React from 'react'
import { applyUid, convertOldStringToDataArray } from './utils'
import './ImportCollection.css';
import { useRecoilValue, useSetRecoilState } from 'recoil';
import { settingsDataState } from './atoms/settingsDataState';
import { rowToHighlightState } from './atoms/animationsState';
import { useSnackbar } from 'react-simple-snackbar';
import { SnackbarStyle } from './model/SnackbarTypes';


function ImportCollection(props) {

    const settingsData = useRecoilValue(settingsDataState);
    const setRowToHighlight = useSetRecoilState(rowToHighlightState);
    const [openSnackbar, closeSnackbar] = useSnackbar({style: SnackbarStyle.ERROR});

    const handleFileSelection = (event) => {
        const file = event.target.files[0];
        if (!event.target.value.endsWith('.txt')) {
            openSnackbar('Invalid file: Please select a .txt file', 4000);
            event.target.value = '';
            return;
        }
        let reader = new FileReader();
        reader.onload = function (e) {
            const result = reader.result;
            if (result.indexOf('`') == -1 && !result.startsWith('{"')) {
                openSnackbar('Invalid File: This file is not a valid collection', 4000);
                event.target.value = '';
                return;
            }
            let newItem;
            if (result.indexOf('`') > -1) {
                newItem = convertOldStringToDataArray(result)[0];
            }
            if (result.startsWith('{"')) {
                newItem = JSON.parse(result);
                newItem['createdOn'] = Date.now();
            }
            if (!('uid' in newItem.tabs[0])) newItem = applyUid(newItem);
            const newData = [newItem, ...settingsData];
            props.updateRemoteData(newData).then(() => {
                setRowToHighlight(0);
                event.target.value = '';
            });
        }
        reader.readAsText(file);

    };
    return <span className="image-upload">
            <label htmlFor="file-input">
            <img src="images/import_file.png" /> <span>Import a collection from file</span>
            </label>
            <input id="file-input" type="file" onChange={handleFileSelection} />
        </span>;
};

export default ImportCollection;