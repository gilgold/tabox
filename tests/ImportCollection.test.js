import React from 'react';
import renderer from 'react-test-renderer';
import ImportCollection from '../app/ImportCollection';
import { RecoilRoot } from 'recoil';

describe('Import Collection section tests', () => {
  test('Import collection component renders correctly', () => {
    const component = renderer.create(
      <RecoilRoot>
        <ImportCollection />
      </RecoilRoot>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});