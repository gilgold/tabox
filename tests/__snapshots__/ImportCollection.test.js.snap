// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Import Collection section tests Import collection component renders correctly 1`] = `
<span
  className="image-upload"
>
  <label
    htmlFor="file-input"
  >
    <img
      src="images/import_file.png"
    />
     
    <span>
      Import a collection from file
    </span>
  </label>
  <input
    id="file-input"
    onChange={[Function]}
    type="file"
  />
</span>
`;
